mava/systems/sac/ff_isac.py

```diff
@@ -0,0 +1,565 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import time
+from pprint import pprint
+from typing import Any, Callable, Dict, Tuple
+
+import chex
+import flashbax as fbx
+import hydra
+import jax
+import jax.lax as lax
+import jax.numpy as jnp
+import optax
+from colorama import Fore, Style
+from flashbax.buffers.flat_buffer import TrajectoryBuffer
+from flax.core.scope import FrozenVariableDict
+from jax import Array
+from jumanji.env import Environment, State
+from omegaconf import DictConfig, OmegaConf
+
+from mava.evaluator import make_eval_fns
+from mava.networks import FeedForwardActor as Actor
+from mava.networks import FeedForwardQNet as QNetwork
+from mava.systems.sac.types import (
+    BufferState,
+    LearnerState,
+    Metrics,
+    Networks,
+    Optimisers,
+    OptStates,
+    Qs,
+    QsAndTarget,
+    SacParams,
+    Transition,
+)
+from mava.types import Observation
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax import unreplicate_batch_dim, unreplicate_n_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.wrappers import episode_metrics
+
+
+def init(
+    cfg: DictConfig,
+) -> Tuple[
+    Tuple[Environment, Environment],
+    Networks,
+    Optimisers,
+    TrajectoryBuffer,
+    LearnerState,
+    Array,
+    MavaLogger,
+    chex.PRNGKey,
+]:
+    logger = MavaLogger(cfg)
+
+    key = jax.random.PRNGKey(cfg.system.seed)
+    devices = jax.devices()
+
+    def replicate(x: Any) -> Any:
+        """First replicate the update batch dim then put on devices."""
+        x = jax.tree_map(lambda y: jnp.broadcast_to(y, (cfg.system.update_batch_size, *y.shape)), x)
+        return jax.device_put_replicated(x, devices)
+
+    env, eval_env = environments.make(cfg)
+
+    n_agents = env.action_spec().shape[0]
+    action_dim = env.action_spec().shape[1]
+
+    key, actor_key, q1_key, q2_key, q1_target_key, q2_target_key = jax.random.split(key, 6)
+
+    acts = env.action_spec().generate_value()  # all agents actions
+    act_single_batched = acts[0][jnp.newaxis, ...]  # batch single agent action
+    obs = env.observation_spec().generate_value()
+    obs_single_batched = jax.tree_map(lambda x: x[0][jnp.newaxis, ...], obs)
+
+    # Making actor network
+    actor_torso = hydra.utils.instantiate(cfg.network.actor_network.pre_torso)
```

Comment:
(Newbie eyes comment) This hydra way of instantiating a network is new to me - how does it know where to look, and does it return the exact same thing as the usual way of instantiating a network?

Reply:
Check in the config, there's a `_target_` key that allows hydra to know where to look to instantiate each config item.

Reply:
Maybe we can add it to the docs somewhere? Can you make a comment or a commit on the [docs PR](https://github.com/instadeepai/Mava/pull/1052) about this please @lbeyers?